### ISSUE TITLE
Periodically send Gateway stats to portal

### ISSUE THREAD

**Issue by jamilbk** [firezone/firezone#2060]
User story:
As an administrator, I'd like to easily know how much data transfer my client devices are consuming in my cloud environment to ensure the data transferred is sane.

<img width="689" alt="Screenshot_2023-09-13_at_7_39_56_AM" src="https://github.com/firezone/firezone/assets/167144/62692bb2-5ca0-4110-be7f-907fdfc288ea">

**Comment by AndrewDryga** [comment]
Portal already has API for this.

**Comment by jamilbk** [comment]
Customer request: add basic system stats as well:

- System load

**Comment by thomaseizinger** [comment]
> Portal already has API for this.

What does the API look like?

**Comment by thomaseizinger** [comment]
Instead of sending it to the portal, should we add metrics to the gateway and export them using OTEL? If we pick https://github.com/firezone/firezone/issues/2033 back up, we could report them to Google Cloud Metrics. Same as for the relay, we can make it configurable in the gateways, where they send the log and telemetry data.

**Comment by jamilbk** [comment]
Haven't had anyone request OTEL yet.

Mostly prometheus and just UI stats summary in the portal so admins and other stakeholders can see a quick glimpse of gateway load.

I think the question users are trying to answer from this data is, "do I need to deploy more gateways?"

**Comment by thomaseizinger** [comment]
> Mostly prometheus

You can get that from OTEL :)

> I think the question users are trying to answer from this data is, "do I need to deploy more gateways?"

So what data do we need for that? Rolling avg bandwidth? Number of clients? The reason I am bringing up OTEL is because it is simpler to implement and self-serve (assuming we capture the right base metrics)! Capturing ingress and egress traffic is quite simple. A user could then calculate whatever avg. bandwidth they'd like! :)

**Comment by jamilbk** [comment]
> So what data do we need for that?

I was thinking of just printing the system load, average throughput, and transfer stats periodically if that's not too complex.

> A user could then calculate whatever avg. bandwidth they'd like! :)

We probably need both. OTEL for the sysadmins / IT folks, and basic UI stats for the other 99% and non-technical stakeholders.

I imagine the value proposition becomes clear when they see the stats in the UI right away for all gateways, but then move to an OTEL collector for production operation. I.e. the pilot / trial workflow -> production setup, kinda like how they can spin up a Gateway via the UI but real production operation will use Terraform.

We've been getting requests for this [for a while](https://github.com/firezone/firezone/issues/540).

**Comment by AndrewDryga** [comment]
@thomaseizinger The gateways are hosted by our customers, so OTEL doesn't fit us here. 

OTEL is just a monitoring tool that we could use if we want to monitor our 2-3 private gateways or if our clients want to monitor theirs, but I don't think there is a point in doing it because data won't be actionable for them and they already can monitor VM/container CPU/MEM usage. (cc @jamilbk, we should ask why they don't just monitor the VM?)

The core of current metrics is to see how much traffic there was, sort of enriched audit logging (eg. you can know if somebody downloaded large files or connected and did nothing at all). The API is [here](https://github.com/firezone/firezone/blob/main/elixir/apps/api/lib/api/gateway/channel.ex#L323):
```
%{
  "started_at" => started_at,
  "ended_at" => ended_at,
  "metrics" => [
    %{
      "flow_id" => flow_id,
      "destination" => destination,
      "rx_bytes" => rx_bytes,
      "tx_bytes" => tx_bytes
    },
    ...
  ]
}
``` 

The `destination` is IP and port (`127.0.0.1:80`) but we can add the protocol to it (eg. `tcp:127.0.0.1:80`).

@jamilbk we could add mem/cpu on top of that but this feels like a very limited feature because what should we do with this information? Just display in the UI? Without alerts, etc it's not that helpful to have it.

**Comment by jamilbk** [comment]
> we could add mem/cpu on top of that but this feels like a very limited feature because what should we do with this information? Just display in the UI? Without alerts, etc it's not that helpful to have it.

Yeah I tend to agree here, better not to add it unless there's a clear action item that can result from this info. CPU and mem monitoring is better done using other infrastructure tooling.

The one thing Firezone can uniquely provide is transfer stats (this should be step 1)

and then network flow logs (including flow 5-tuples: `proto:src:src_port:dst:dst_port`) from the decapsulated packet.

**Comment by thomaseizinger** [comment]
> I don't think there is a point in doing it because data won't be actionable for them and they already can monitor VM/container CPU/MEM usage

Hmm, I am surprised about that view. I would have thought that OTEL-based stats are much more actionable to users because they can use existing tooling and are not limited by whatever visualisation we provide in the portal.

> and then network flow logs (including flow 5-tuples: `proto:src:src_port:dst:dst_port`) from the decapsulated packet.

A thought regarding that: Prometheus strongly discourages collecting metrics that are categorised by IPs, ports or anything else that is not bounded to a small set because it will lead to a cardinality explosion. I am assuming the same applies to every system that wants to store metrics as a time-series.

I think the conclusion from this is: Yes we can expose basic metrics in the portal based on 5-tuples if we pre-aggregate them but it also means users are limited to that aggregation of the data. Anything that we expose as prometheus or OTEL metrics cannot be as fine-granular if we don't want to DoS whatever time-series DB users are using to store them.

Metrics that we could easily expose are:

- Protocol category (UDP, TCP, ICMP, etc)
- total bytes sent & received
- total bytes written and read from network (allows computing overhead by STUN, TURN and WG)
- Traffic category (direct-direct, direct-relay, relay-direct, relay-relay)
- number of active connections
- histogram of packet-switch latency (i.e. time from reading a packet to sending it out to the device)

**Comment by AndrewDryga** [comment]
> Hmm, I am surprised about that view. I would have thought that OTEL-based stats are much more actionable to users because they can use existing tooling and are not limited by whatever visualisation we provide in the portal.

I meant there are existing agents you can install on the host (whatever it is) and send CPU/RAM usage to OTEL or any other tool without us implementing anything if the goal is to monitor the host to prevent overloads. This is mostly for DevOps people to alert on scaling issues.

I guess it's not bad if we send those to the portal too, we could send emails to admins when gateways are overloaded, but this feels like the limit of what we should do before we turn into a monitoring tool.

> I think the conclusion from this is: Yes we can expose basic metrics in the portal based on 5-tuples if we pre-aggregate them but it also means users are limited to that aggregation of the data. Anything that we expose as prometheus or OTEL metrics cannot be as fine-granular if we don't want to DoS whatever time-series DB users are using to store them.

I think there is a slight confusion here, I believe what @jamilbk wants as flow logs is the thing that the portal accepts over websocket. In the future, we could stream 5-tuples to syslog so that our clients can build their own activity logs (for security) and ingest them to something like Splunk, but this is not for OTEL.

**Comment by jamilbk** [comment]
Sorry yeah I should have clarified -- Netflows are usually aggregated. We can discuss more about netflow streaming into SIEM when we have a specific request for it.

These are usually for creating workflows that intent to find indicators of compromise by correlating logs from various sources.

An example workflow an incident response team might use is:

- Stream netflow / access logs into Splunk
- If the request is:
  - detected regularly to a specified resource (dest IP + port),
  - by a certain device platform (iOS 16.3.1),
  - from a certain region or user,
  - of a certain size or duration,

Then that might provide a high confidence IoC for a specific type of malware that would be hard to tie together without Firezone's help

**Comment by jamilbk** [comment]
> the thing that the portal accepts over websocket.

TBD. If this forces the customer to pay egress fees in their cloud and then us pay egress fees back to the customer's cloud Splunk collector maybe the architecture could be direct streaming instead -- I think that is what other products do for high volume streaming.

For example, Splunk has a GCP SaaS offering, example setup: https://cloud.google.com/architecture/stream-logs-from-google-cloud-to-splunk

They probably have a marketplace SaaS app for all major clouds to avoid egress

**Comment by AndrewDryga** [comment]
@jamilbk there won't be high fees for aggregated logs streamed from the portal, it's only expensive if you stream them for every packet

**Comment by jamilbk** [comment]
I think it's worth revisiting this feature, especially for #4929. The portal handles the aggregations right? So we just need a timer-based mechanism from the Gateway to report TX/RX for each flow_id, right?

**Comment by jamilbk** [comment]
Addressed with the introduction of #14443
